pyramid_swagger/tween.py

```diff
@@ -103,21 +108,89 @@ def validator_tween(request):
                 return handler(request)
 
         if settings.validate_request:
-            with validation_context(request):
-                route_match = (route_info.get('match') or {}).items()
-                _validate_request(route_match, request, validator_map)
+            request_data = handle_request(
+                PyramidSwaggerRequest(request, route_info),
+                validation_context,
+                validator_map)
+
+            def swagger_data(_):
+                return request_data
+
+            request.set_property(swagger_data)
 
         response = handler(request)
 
         if settings.validate_response:
             with validation_context(request, response=response):
-                _validate_response(response, validator_map.response)
+                validate_response(response, validator_map.response)
 
         return response
 
     return validator_tween
 
 
+class PyramidSwaggerRequest(object):
+    """Adapter for a :class:`pyramid.request.Request` which exposes request
+    data for casting and validation.
+    """
+
+    def __init__(self, request, route_info):
+        self.request = request
+        self.route_info = route_info
+
+    @property
+    def query(self):
+        return self.request.GET
+
+    @property
+    def path(self):
+        return self.route_info.get('match') or {}
+
+    @property
+    def headers(self):
+        return self.request.headers
+
+    @property
+    def body(self):
+        return getattr(self.request, 'json_body', {})
+
+
+def handle_request(request, validation_context, validator_map):
+    """Validate the request against the swagger spec and return a dict with
+    all parameter values available in the request, casted to the expected
+    python type.
+
+    :param request: a :class:`PyramidSwaggerRequest` to validate
+    :param validation_context: a context manager for wrapping validation
+        errors
+    :param validator_map: a :class:`pyramid_swagger.load_schema.ValidatorMap`
+        used to validate the request
+    :returns: a :class:`dict` of request data for each parameter in the swagger
+        spec
+    """
+    request_data = {}
+    validation_pairs = []
+
+    for validator, values in [
+        (validator_map.query, request.query),
+        (validator_map.path, request.path),
+        (validator_map.headers, request.headers),
+    ]:
+        values = cast_params(validator.schema, values)
+        validation_pairs.append((validator, values))
+        request_data.update(values)
+
+    if validator_map.body.schema:
```

Comment:
I agree. I really tried to get this inline, but in the end it's special-cased because the name is external to the body itself, where as the other `values` contain the key in the dict.

I should add a comment about that